### PR TITLE
Fix #7464 stripe

### LINKED ISSUE #7464: add stripe payment option

### PR DESCRIPTION
**Pull request by e-carlin**
Add ability to pay for basic/premium plan with Stripe.

There are still some open TODO's that we can discuss in the review.
- [ ] How does this work with canceled due to timeout? If they are on trial we probably want to show upgrade to premium not upgrade to basic (basic has same run time). If they are basic we'll want to show upgrade to premium.
- [ ] simulation_db import in payments_test
- [ ] Has the WP developer added buttons from the plans page? They'll need links from us
- [ ] The "subscribe now" link always goes to basic. It should probably go to the plans page and let the user select basic or premium.
- [ ] have to put auth_db on subscription record for save to work (see auth_db/user.py)
- [ ] 

I also need to get Stripe completely sorted out
- [ ] Remove payment options like amazon (just have credit card payment)
- [ ] Remove "1 click payment with link"
- [ ] Prevent users from having multiple of the same subscription
- [ ] Add icons
- [ ] Copy test products (basic+premium) to prod
- [ ] Figure out taxes (do we need to collect tax? should stripe handle?)
- [ ] Add our API endpoint for Stripe. Just prod is probably good enough. I think we should have beta point to Stripe prod so we can test for real before prod.

Rsconf
- [ ] configure Sirepo with Stripe keys/prices

@robnagler @moellep please test. There are a lot of edge cases so play around with your roles to test (have an active trial and upgrade to basic, have an expired trial and upgrade, have an expired trial and basic and try to use sirepo, etc)

### CODE REVIEW

**Comment by robnagler** [comment]
@e-carlin I'll leave these to you (or @moellep):
- [ ] "Plan Required" should be more verbose
- [ ] comes up too many times on simulation page

Screenshot 2025-03-17 at 17 48 54

Screenshot 2025-03-17 at 17 48 17

**Comment by e-carlin** [comment]
@moellep would you mind taking on those issues?

I'm going to fix the tests.

**Comment by robnagler** [comment]
Just a note to future self:

`sirepo roles add_or_update vagrant@localhost.localdomain trial --expiration=-1`

This allows you to expire an existing subscription from the command line immediately.

**Review by robnagler — APPROVED** [review]
I'm not going to block the release for these changes, but I think we should discuss if they should be made or not.

**Comment by robnagler** [comment]
More language issues on the page where you click on a sim without a plan:

# Plan Required
An [active plan](#) is required to use Sirepo.

"plan" is a word we use internally. The external word should be subscription. The alert above this message says:

**Your Sirepo free trial has expired.** [Subscribe now](#)

So

# Subscription Required

[Click here to subscribe](#)

**Comment by robnagler** [comment]
Not for now, but noting. You can click on Open as New Copy, New Simulation, New Folder, Import, which prompts you and then when you submit, it shows plan required.

**Comment by robnagler** [comment]
# Payment Complete
Thank you for signing up! Please [click here](#7474 ) to use Sirepo.

Change this to:

# Thank you for your Payment
[Click here for to use start using Sirepo!](#)

File: sirepo/auth_db/user.py [hunk 1]
```diff
@@ -13,6 +13,27 @@
 import sqlalchemy
 
 
+class UserPayment(sirepo.auth_db.UserDbBase):
+    __tablename__ = "user_payment_t"
+    # invoice_id's are unique so safe for a primary_key and makes
+    # payment_exists check easier.
+    # https://docs.stripe.com/api/invoices/object#invoice_object-id
+    invoice_id = sqlalchemy.Column(sirepo.auth_db.STRING_NAME, primary_key=True)
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/auth_db/user.py]: 94ffb4963

File: sirepo/auth_db/user.py [hunk 2]
```diff
@@ -13,6 +13,27 @@
...
+    uid = sqlalchemy.Column(sirepo.auth_db.STRING_ID, nullable=False)
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/auth_db/user.py:24]: Discussed skipping foreign keys and delete for now.

File: sirepo/auth_db/user.py [hunk 3]
```diff
@@ -156,3 +198,40 @@ def set_status(self, role, status, moderator_uid):
...
+
+
+class UserSubscription(sirepo.auth_db.UserDbBase):
+    CREATION_REASON_CHECKOUT_SESSION_STATUS_COMPLETE = (
+        "payments_checkout_session_status_complete"
+    )
+    _REVOCATION_REASON_INACTIVE_STRIPE_STATUS = "inactive_stripe_status"
+
+    __tablename__ = "user_subscription_t"
+    id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/auth_db/user.py]: How does this look? 94ffb4963

File: sirepo/auth/__init__.py [hunk 4]
```diff
@@ -679,15 +673,15 @@ def _auth_state(self):
         )
         if "sbatch" in v.jobRunModeMap:
             v.sbatchQueueMaxes = sirepo.job.NERSC_QUEUE_MAX
+        if "payments" in sirepo.feature_config.cfg().api_modules:
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/auth/__init__.py]: 6105cc041

File: sirepo/auth/__init__.py [hunk 5]
```diff
@@ -696,7 +690,10 @@ def _auth_state(self):
                 )
                 if r:
                     v.displayName = r.display_name
-            v.roles = self.qcall.auth_db.model("UserRole").get_roles()
+            v.roles = {
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/auth/__init__.py:694]: Who should do the timestamp conversion?

File: sirepo/auth_role_moderation.py [hunk 6]
```diff
@@ -137,15 +136,10 @@ def _set_moderation_status(info):
             _send_moderation_status_email(p)
         return self.reply_ok()
 
-    @sirepo.quest.Spec("require_adm")
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/auth_role_moderation.py:140]: W/o this change if you are not logged in then you are taken to /. With this change you are taken to the login page which seems like a more appropriate location.

File: sirepo/util.py [hunk 7]
```diff
@@ -293,6 +293,15 @@ def find_obj(arr, key, value):
     return None
 
 
+def first_sim_type():
```
[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/util.py:298]: f786eedfb

**Comment by e-carlin** [comment]
@robnagler how did you get this to come up https://github.com/radiasoft/sirepo/pull/7474#issuecomment-2731211823 ? I always get taken to the plan required page.

**Comment by robnagler** [comment]
> how did you get this to come up https://github.com/radiasoft/sirepo/pull/7474#issuecomment-2731211823 ? I always get taken to the plan required page.

I was in a simulation already in another window(?) or I expired the subscription in a simulation window.

**Comment by moellep** [comment]
In 591aa04 I changed it to redirect to the plan page for these errors.

**Comment by robnagler** [comment]
This is caused by #7487

File: sirepo/auth/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/auth/__init__.py:694]: get_roles_and_expiration.

File: sirepo/auth/email.py [hunk 8]
```diff
@@ -67,27 +66,32 @@ def _verify_confirm(sim_type, user):
...
-                display_name=n,
-            )
+            try:
+                self.auth.login(
+                    this_module,
+                    sim_type=req.type,
+                    model=u,
+                    display_name=d.get("displayName"),
+                )
+            except sirepo.util.SReplyExc:
```
[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/auth/email.py:90]: I don't want to add moderate to login. That's already too complicated as it is. I think the code is fine, just looking at the exception.

File: sirepo/auth_role_moderation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/auth_role_moderation.py:140]: This is for adm. You should be logged in. If you are not logged in, it should go to a forbidden page, not "/", since the end user is a sophisticated user. The forbidden page could look at the cookie and redirect to login or not. That's how BOP works. If you see there's a logged out uid, you simply redirect to login page. This is an unlikely case in this specific code. Changing the permission here is a potential security issue. adm tasks (not matter what) should always require_adm.

**Review by robnagler — APPROVED** [review]
The only change that is important is the require_adm, which I just did. We need to move on at this point.
